software/contrib/master_clock.py

```diff
@@ -134,10 +149,71 @@ def cycleScreen():
                 else:
                     self.screen = 1
 
-        # Reset step count upon receiving a din voltage
+        # Trigger clock if using an external clock, or reset if not
         @din.handler
-        def resetDin():
-            self.step = 1
+        def dinTrigger():
+            if self.externalClockInput:
+                # Divide input clocks by self.inputClockDivision and trigger the clock
+                if self.clockInputNum % self.inputClockDivision == 0:
+                    self.clockTrigger()
+                    if self.clockInputNum > 1: # Ignore the first entry as it has no reference
+                        self.mSBetweenClockCycles = time.ticks_diff(ticks_ms(), self.previousClockTime)
+                        # Cap the BPM calcs at at 240 BPM to avoid steam coming out of the pico's ears
+                        if self.mSBetweenClockCycles < 250:
+                            self.inputClockDiffs.append(self.mSBetweenClockCycles)
+                        else:
+                            self.inputClockDiffs.append(250)
+                        # Only keep 20 values in the buffer
+                        if len(self.inputClockDiffs) == 20:
+                            del self.inputClockDiffs[0]
+
+                        if self.clockInputNum > 3: # Only calculate is there are > 3 entries
+                            self.bpm = self.calculateBpm(self.inputClockDiffs)
+                    self.previousClockTime = ticks_ms()
+                    
+                self.clockInputNum += 1
+            else:
+                self.step = 1
+
+    ''' Ask to use internal or external clock on boot'''
+    def getClockOption(self):
+        self.clockSelectionScreenActive = True
+        oled.fill(0)
+        oled.text("Clock Source:", 0, 0, 1)
+        oled.text("B1: Internal", 0, 9, 1)
+        oled.text("B2: External", 0, 17, 1)
+        oled.show()
+        while True:
+            if b1.value() == 1:
+                self.externalClockInput = False
+                self.running = False # Need to do this to keep it running because the b1 handler will reverse the value
+                self.clockSelectionScreenActive = False
+                break
+            elif b2.value() == 1:
+                self.externalClockInput = True
+                self.clockSelectionScreenActive = False
+                break
+            time.sleep(0.05)
+        
+        self.saveState()
+
+    def median(self, lst):
+        n = len(lst)
+        s = sorted(lst)
+        return (s[n//2-1]/2.0+s[n//2]/2.0, s[n//2])[n % 2] if n else None
+
+    def bpmFromMs(self, ms):
+        return int(((1/(ms/1000))*60)/4)
+
+    def calculateBpm(self, list):
+        # Create a copy of the list, so the sort command does not affect the original list
+        lst = list[:]
+        lst.sort()
+        self.averageDiff = self.average(list)
```

Comment:
[question] Why do you need to sort the list to get its average?

Reply:
Good spot. This code if left over from when I had logic that did a mean or median average - the latter needed the sort. The median is no longer needed, so I will remove the median() function and also the list sorting code.